test/jdk/javax/swing/JToolBar/ToolBarSeparatorTest.java

```diff
@@ -0,0 +1,108 @@
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 4346610
+ * @key headful
+ * @summary Verifies if Adding JSeparator to JToolBar "pushes" buttons added
+ *          after separator to edge
+ * @run main ToolBarSeparatorTest
+ */
+import java.awt.BorderLayout;
+import java.awt.Color;
+import java.awt.Dimension;
+import java.awt.Insets;
+import java.awt.image.BufferedImage;
+import java.awt.Point;
+import java.awt.Rectangle;
+import java.awt.Robot;
+
+import javax.swing.JButton;
+import javax.swing.JFrame;
+import javax.swing.JPanel;
+import javax.swing.JSeparator;
+import javax.swing.JToolBar;
+import javax.swing.SwingConstants;
+import javax.swing.SwingUtilities;
+
+import javax.imageio.ImageIO;
+
+public class ToolBarSeparatorTest {
+
+    private static JFrame frame;
+    private static JSeparator separator;
+    private static JToolBar toolBar;
+    private static JButton btn;
+    private static volatile Point pt;
+    private static volatile Dimension size;
+    private static volatile int sepWidth;
+    private static volatile int sepPrefWidth;
+
+    public static void main(String[] args) throws Exception {
+        Robot robot = new Robot();
+        robot.setAutoDelay(100);
+        try {
+            SwingUtilities.invokeAndWait(() -> {
+                frame = new JFrame("Troy's ToolBarTest");
+                toolBar = new JToolBar();
+                toolBar.setMargin(new Insets(0,0,0,0));
+                btn = new JButton("button 1");
+                toolBar.add(btn);
+                toolBar.add(new JButton("button 2"));
+                separator = new JSeparator(SwingConstants.VERTICAL);
+                toolBar.add(separator);
+                toolBar.add(new JButton("button 3"));
+                toolBar.setBackground(Color.red);
+                frame.getContentPane().setLayout(new BorderLayout());
+                frame.getContentPane().add(toolBar, BorderLayout.NORTH);
+                frame.getContentPane().add(new JPanel(), BorderLayout.CENTER);
+                frame.setSize(400, 100);
+                frame.setLocationRelativeTo(null);
+                frame.setVisible(true);
+            });
+            robot.waitForIdle();
+            robot.delay(1000);
+            SwingUtilities.invokeAndWait(() -> {
+                pt = toolBar.getLocationOnScreen();
+                size = toolBar.getSize();
+                sepWidth = separator.getSize().width;
+                sepPrefWidth = separator.getPreferredSize().width;
+            });
+            if (separator.getSize().width != separator.getPreferredSize().width) {
+                System.out.println("size " + sepWidth);
+                System.out.println("preferredsize " + sepPrefWidth);
+                BufferedImage img = robot.createScreenCapture(
+                    new Rectangle(pt.x, pt.y, size.width, size.height));
+                ImageIO.write(img, "png", new java.io.File("image.png"));
```

Comment:
Theoretically, though unlikely, it may throw `IOException` which will hide the following `RuntimeException` which fails the test. But it's okay: the test fails one way or another.